### PR TITLE
feat: update checker

### PR DESCRIPTION
**Pull request by crowlKats**

### CODE REVIEW

**Comment by kitsonk** [comment]
You don't want this running under the lsp subcommand, as the output would be missed.  I assume there are other subcommands it should be supressed.

Also, it seems to check every invocation?

File: cli/main.rs [hunk 1]
```diff
@@ -1071,10 +1071,67 @@ pub fn main() {
...
+    if let Some(ca_file) = ca_file {
+      let buf = std::fs::read(ca_file)?;
+      let cert = reqwest::Certificate::from_pem(&buf)?;
+      client_builder = client_builder.add_root_certificate(cert);
+    }
+
+    let client = client_builder.build()?;
+
+    let ver = if version::is_canary() {
+      let res = client.get("https://version.deno.dev/canary").send().await?;
```
Comment (kitsonk) [inline cli/main.rs:1103]: hardcoded values "buried" in code feels bad... maybe should be extracted out as a const?

Comment (kitsonk) [inline cli/main.rs:1103]: shouldn't we be setting the UA string like with the file fetcher?

**Comment by crowlKats** [comment]
yes, it shouldnt run under some subcommands, and yes, it shouldnt run on every invocation; PR is not ready yet

**Comment by kitsonk** [comment]
> yes, it shouldnt run under some subcommands, and yes, it shouldnt run on every invocation; PR is not ready yet

ok, let me know when you want me to complain

**Review by littledivy — COMMENTED** [review]
im concerned about startup time. how bad is it compared to `main`?

**Comment by piscisaureus** [comment]
To avoid ruining startup time, I think we should do the check in the background and display update info on the next invocation.

File: cli/main.rs [hunk 2]
```diff
@@ -1071,10 +1071,67 @@ pub fn main() {
...
+    if let Some(ver) = ver {
+      if !std::fs::metadata(env::current_exe()?)?
+        .permissions()
+        .readonly() {
+        println!(
+          "{}",
+          colors::green_bold(format!(
+            "Version '{ver}' of Deno is out, update with `deno upgrade`."
+          ))
+        );
```
Comment (dsherret) [inline cli/main.rs:1131]: I was just thinking that we should maybe only output this when stderr is a tty (and this outputs to stderr instead). That way we won’t break anyone piping deno process output randomly.

File: cli/main.rs (hunk 1 above)
Comment (piscisaureus) [reply cli/main.rs:1103]: We can just use https://dl.deno.land/release-latest.txt and https://dl.deno.land/canary-latest.txt for this.

File: cli/main.rs (hunk 2 above)
Comment (piscisaureus) [reply cli/main.rs:1131]: Agree! Good idea.

**Comment by bartlomieju** [comment]
Closing in favor of https://github.com/denoland/deno/pull/15974

**Pull request closed without merging**